cachito/workers/pkg_managers/yarn.py

```diff
@@ -90,10 +115,24 @@ def _get_deps(yarn_lock, file_deps_allowlist):
     """
     deps = []
     nexus_replacements = {}
+    non_dev_deps = set()
+
+    for dep_type in ["dependencies", "peerDependencies", "optionalDependencies"]:
+        if dep_type not in package_json:
+            continue
+        for name, version in package_json[dep_type].items():
+            dep = f"{name}@{version}"
+            if dep not in non_dev_deps:
+                _find_reachable_deps(yarn_lock, dep, non_dev_deps)
 
     for dep_identifier, dep_data in yarn_lock.items():
         package = pyarn.lockfile.Package.from_dict(dep_identifier, dep_data)
 
+        dev = True
+        for non_dev_dep in non_dev_deps:
+            if non_dev_dep in dep_identifier:
+                dev = False
```

Comment:
Is substring matching reliable enough? What about cases like this:
```python
>>> non_dev_deps = {"react@1.0.0"}
>>> "react@1.0.0" in "@types/react@1.0.0"
True
```

Reply:
I made a helper function `_split_lock_key` which splits the `dep_identifier` string into a list of dependencies (removes unnecessary " and splits string by ", "). WDYT?

i.e.
```
>>> dep_id_list = _split_lock_key('"@types/react@1.0.0, @types/react@1.2.0"')
>>> dep_id_list
["@types/react@1.0.0", "@types/react@1.2.0"]
>>> "react@1.0.0" in dep_id_list
False
>>> "@types/react@1.0.0" in dep_id_list
True
```

Reply:
+1, that should work. How about using the helper function for `_expand_replacements` as well, to make sure they do this consistently? The current approach is to split on `","` and then strip whitespace, which might be slightly more reliable - but there is no handling for extra quotes, which would be good to have.

As a side note, `_expand_replacements` may deserve a name change, it is now used to expand the entire  yarn.lock. Maybe `_expand_yarn_lock_keys`?

Reply:
And maybe, instead of iterating through all `non_dev_deps` and checking if they are in `dep_id_list`, would it be better to do it the other way around? The dep_id_list will usually have only one or two items, and since non_dev_deps is a set that should be constant lookup on average.